{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "Collapsed": "false"
   },
   "source": [
    "https://pytorch.org/tutorials/beginner/nn_tutorial.html\n",
    "\n",
    "https://github.com/spro/practical-pytorch/tree/master/char-rnn-classification\n",
    "\n",
    "https://pytorch.org/tutorials/beginner/text_sentiment_ngrams_tutorial.html\n",
    "\n",
    "Batch RNN\n",
    "https://pytorch.org/tutorials/beginner/text_sentiment_ngrams_tutorial.html\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "Collapsed": "false"
   },
   "source": [
    "![alt text](https://docs.gdc.cancer.gov/Encyclopedia/pages/images/barcode.png \"TCGA Barcode\")\n",
    "\n",
    "To identify a sample, which is used for both methylation and expression measurements, we need only first 4 \"-\" separated identifiers, i.e. up to sample:vial level;\n",
    "\n",
    "It is assumed that each patient is measured once (valid)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [
    {
     "ename": "ImportError",
     "evalue": "TensorBoard logging requires TensorBoard with Python summary writer installed. This should be available in 1.14 or above.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/envs/ortholog/lib/python3.7/site-packages/torch/utils/tensorboard/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0;32mfrom\u001b[0m \u001b[0mtensorboard\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msummary\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwriter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecord_writer\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mRecordWriter\u001b[0m  \u001b[0;31m# noqa F401\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;32mexcept\u001b[0m \u001b[0mImportError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'tensorboard'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-c1342485bfe8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mtorch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptim\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0moptim\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mtorch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mutils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mDataset\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mDataLoader\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mtorch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mutils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtensorboard\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mSummaryWriter\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0mtorch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmanual_seed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m42\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/ortholog/lib/python3.7/site-packages/torch/utils/tensorboard/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0;32mfrom\u001b[0m \u001b[0mtensorboard\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msummary\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwriter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecord_writer\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mRecordWriter\u001b[0m  \u001b[0;31m# noqa F401\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mexcept\u001b[0m \u001b[0mImportError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     raise ImportError('TensorBoard logging requires TensorBoard with Python summary writer installed. '\n\u001b[0m\u001b[1;32m      5\u001b[0m                       'This should be available in 1.14 or above.')\n\u001b[1;32m      6\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mwriter\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mFileWriter\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSummaryWriter\u001b[0m  \u001b[0;31m# noqa F401\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mImportError\u001b[0m: TensorBoard logging requires TensorBoard with Python summary writer installed. This should be available in 1.14 or above."
     ]
    }
   ],
   "source": [
    "import re\n",
    "import numpy as np\n",
    "\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "import torch.optim as optim\n",
    "from torch.utils.data import Dataset, DataLoader\n",
    "from torch.utils.tensorboard import SummaryWriter\n",
    "torch.manual_seed(42)\n",
    "\n",
    "\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "\n",
    "DIR = Path(\"/data/eugen/meth2expr/\")\n",
    "TCGA_DIR = Path(\"/data/eugen/tcga/projects\")\n",
    "\n",
    "projects = [str(x).split(\"/\")[-1] for x in TCGA_DIR.iterdir() if x.is_dir()]\n",
    "projects_paths = [x for x in TCGA_DIR.iterdir() if x.is_dir()]\n",
    "\n",
    "\n",
    "\n",
    "#print(torch.cuda.is_available())\n",
    "#if torch.cuda.is_available() or True:\n",
    "#    device = torch.device(\"cuda\")          # a CUDA device object\n",
    "#    y = torch.ones_like(x, device=device)  # directly create a tensor on GPU\n",
    "#    x = x.to(device)                       # or just use strings ``.to(\"cuda\")``\n",
    "#    z = x + y\n",
    "#    print(z)\n",
    "#    print(z.to(\"cpu\", torch.double))   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": [
    "PROJECTS_DIR = \"/data/eugen/tcga/projects/\"\n",
    "\n",
    "class Project(object):\n",
    "    \"\"\"\n",
    "    Project data loader and descriptor.\n",
    "    \n",
    "    :attributes:\n",
    "    \n",
    "    :public:\n",
    "        get_case_data()\n",
    "    \n",
    "    :private:\n",
    "        _collect_samples()\n",
    "        _collect_metadata()\n",
    "        _get_case_datapaths()\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, name, projects_dir=PROJECTS_DIR):\n",
    "        '''\n",
    "        :params\n",
    "            name -- project name\n",
    "            path -- project dir\n",
    "        '''\n",
    "        self.name = name\n",
    "        self.dir = projects_dir + name\n",
    "        \n",
    "        self.meth_path = None\n",
    "        self.meth_fpath = None\n",
    "        \n",
    "        self.samples = {}\n",
    "        self.sample_ids = []\n",
    "        self.sample_paths = {}\n",
    "        self.cases = {}\n",
    "        self.case_ids = []\n",
    "        self.meta = {}\n",
    "        \n",
    "        # Get samples ids and paths\n",
    "        self._collect_samples()\n",
    "        self._collect_metadata()\n",
    "        \n",
    "        \n",
    "    def _collect_samples(self):\n",
    "        \"\"\"\n",
    "        Extracts samples' ids and paths from project directory.\n",
    "        \"\"\"\n",
    "        \n",
    "        self.sample_paths = {\"methylation\": {},\n",
    "                              \"expression\": {}}\n",
    "        \n",
    "        # Extract methylation\n",
    "        self.meth_path = Path(self.dir) / 'data/methylation' / self.name / 'harmonized/DNA_Methylation/Methylation_Beta_Value'\n",
    "        self.expr_path = Path(self.dir) / 'data/expression' / self.name / 'harmonized/Transcriptome_Profiling/Gene_Expression_Quantification'\n",
    "        self.sample_paths = {datatype: {str(f).split(\"/\")[-1]: str(f) \n",
    "                                        for f in path.iterdir()}\n",
    "                             for datatype, path in [('methylation', self.meth_path), ('expression', self.expr_path)]}\n",
    "        \n",
    "        # In path there is an experiment ID, not sample id\n",
    "        self.samples = {f: list(self.sample_paths[f].keys()) for f in ['methylation', 'expression']}\n",
    "        self.sample_ids = list(self.samples.keys())\n",
    "    \n",
    "    \n",
    "    def _collect_metadata(self):\n",
    "        \"\"\"\n",
    "        Gets all necessary metadata of project.\n",
    "        Includes ids and case ids: necessary to match expression and methylation measurements.\n",
    "        \"\"\"\n",
    "        \n",
    "        metadata_paths = [(Path(self.dir) / (self.name + \"_\" + f + \".csv\"))\n",
    "                              for f in ['methylation', 'expression'] ]\n",
    "        \n",
    "        ### Access metadata files\n",
    "        cols = ['file_id', 'file_name', 'cases']\n",
    "        meth, expr = list(map(lambda p: pd.read_csv(p, sep='\\t', usecols=cols), metadata_paths))\n",
    "        \n",
    "        \n",
    "        ### Prune the TCGA barcode of case ids up to sample level, resulting in Project-TSS-ParticipantID-<SampleID><vial>\n",
    "        barcode_splitter = lambda x: \"-\".join(x.split('-')[:4])\n",
    "        meth['cases'] = meth['cases'].apply(barcode_splitter)\n",
    "        expr['cases'] = expr['cases'].apply(barcode_splitter)\n",
    "\n",
    "                          \n",
    "        meth = meth.set_index(\"file_id\", drop=False)\n",
    "        expr = expr.set_index(\"file_id\", drop=False)\n",
    "        \n",
    "        meta = {'methylation': meth.to_dict(orient='index'),\n",
    "                'expression': expr.to_dict(orient='index') }\n",
    "        \n",
    "        # Assert equivalency of ids in metadata files and available files\n",
    "        assert set(meta['methylation'].keys()) == set(self.samples['methylation'])\n",
    "        assert set(meta['expression'].keys()) == set(self.samples['expression'])\n",
    "        \n",
    "        self.meta = meta\n",
    "        \n",
    "        \n",
    "        ### Collect metadata for cases\n",
    "        meth = meth.set_index(\"cases\")\n",
    "        expr = expr.set_index(\"cases\")\n",
    "        \n",
    "        cases = {'methylation': meth.to_dict(orient='index'),\n",
    "                 'expression': expr.to_dict(orient='index') }\n",
    "        self.old_cases = cases\n",
    "        # recheck equivalency of cases for methylation and expression (should be true by constraint of download script)\n",
    "        assert set(cases['methylation'].keys()) == set(cases['expression'].keys())\n",
    "        \n",
    "        # make cases as keys { case -> {methylation -> {}; expression  -> {}} }\n",
    "        new_dict = {case: {'methylation': cases['methylation'][case],\n",
    "                           'expression': cases['expression'][case]}\n",
    "                    for case in cases['methylation'].keys() }\n",
    "        \n",
    "        self.cases = new_dict\n",
    "        self.case_ids = list(self.cases.keys())\n",
    "        \n",
    "        \n",
    "    def _get_case_datapaths(self, case: str):\n",
    "        '''Getter method for meth, expr or both types data.'''\n",
    "        \n",
    "        assert case in self.cases.keys()\n",
    "        \n",
    "        self.meth_fpath = Path(self.meth_path) / self.cases[case]['methylation']['file_id'] / self.cases[case]['methylation']['file_name']\n",
    "        self.expr_fpath = Path(self.expr_path) / self.cases[case]['expression']['file_id'] / self.cases[case]['expression']['file_name']\n",
    "        \n",
    "        return {'methylation': self.meth_fpath, 'expression': self.expr_fpath}\n",
    "    \n",
    "    \n",
    "    def get_case_data(self, case: str, genes = None, cpgs = None) -> dict:\n",
    "        \"\"\"\n",
    "        Get case methylation and expression dataframe.\n",
    "        \n",
    "        :params\n",
    "            case -- case id;\n",
    "            genes -- list of genes to get data for. if None: gets data for all genes;\n",
    "            cpgs -- list of cpgs to get data for. if None: gets data for all cpgs;\n",
    "        \"\"\"\n",
    "        \n",
    "        get_expr, get_meth = True, True\n",
    "        paths = self._get_case_datapaths(case)\n",
    "        \n",
    "        # We ignore rest columns, since terribly redundant. They describe CpG features, which are assumed to be constant for every\n",
    "        # experiment (to be tested). That's the reason for the terribly inefficient data storage. After testing invariance,\n",
    "        # methylation data files have to cleared (i.e. remove redundant columns).\n",
    "        meth_cols = ['Beta_value']  # 'Composite Element REF' becomes index\n",
    "        \n",
    "        # load both meth and expr for now\n",
    "        # if dtype == 'methylation':\n",
    "        #     get_expr = False\n",
    "        # elif dtype == 'expression':\n",
    "        #     get_meth = False\n",
    "        \n",
    "        if get_expr:\n",
    "            if genes is not None:\n",
    "                expr = pd.read_csv(paths['expression'], sep='\\t', index_col=0, header=None).loc[genes]\n",
    "            else:\n",
    "                expr = pd.read_csv(paths['expression'], sep='\\t', index_col=0, header=None)\n",
    "        \n",
    "        if get_meth:\n",
    "            if cpgs is not None:\n",
    "                meth = pd.read_csv(paths['methylation'], sep='\\t', index_col=0, header=0)[meth_cols].loc[cpgs]\n",
    "            else:\n",
    "                meth = pd.read_csv(paths['methylation'], sep='\\t', index_col=0, header=0)[meth_cols]\n",
    "        \n",
    "        return {'methylation': meth, 'expression': expr}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": [
    "def base2tensor(base: str, vocab: list = VOCAB) -> torch.tensor:\n",
    "    tensor = torch.zeros(1, len(vocab))\n",
    "    tensor[BASE2IDX[base]][0] = 1\n",
    "    return tensor\n",
    "    \n",
    "def seq2tensor(seq: str, vocab: list = VOCAB) -> torch.tensor:\n",
    "    tensor = torch.zeros(len(seq), 1, len(vocab))  # important choice for preserving shape compatibility with hidden layer.\n",
    "    for idx, base in enumerate(seq):\n",
    "        tensor[idx][0][BASE2IDX[base]] = 1  # that extra 1 dimension is because PyTorch assumes everything is in batches - we’re just using a batch size of 1 here.\n",
    "    return tensor\n",
    "\n",
    "def methylate_seq(seq_tensor: torch.tensor, loc: int, value: float, meth_idx: int, mask_idx: int) -> torch.tensor:\n",
    "    assert (seq_tensor[loc][0][meth_idx] > 0) or (seq_tensor[loc][0][mask_idx] > 0) # assert it is a \"C\" in first place\n",
    "    seq_tensor[loc][0][meth_idx] = value\n",
    "    return seq_tensor\n",
    "\n",
    "#################\n",
    "\n",
    "class Gene(object):\n",
    "    \"\"\"\n",
    "    .\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, name, genomics_dir):\n",
    "        \"\"\"\n",
    "        TODO: think about how to compute for each gene local coordinates for CpGs within promoter.\n",
    "        \"\"\"\n",
    "        self.dir = genomics_dir\n",
    "        self.name = name\n",
    "        self.cpgs = []\n",
    "        self.promoter_seq = None  # promoter sequence? or coordinates, etc.\n",
    "        self.tensor = None\n",
    "        \n",
    "        self.mask_idx = -1\n",
    "        self.meth_idx = -1\n",
    "        \n",
    "    def get_promoter_seq(self):\n",
    "        \"\"\"\n",
    "        Load the promoter sequence.\n",
    "        \"\"\"\n",
    "        raise NotImplementedError\n",
    "    \n",
    "    def get_cpgs(self):\n",
    "        \"\"\"\n",
    "        Load the gene's cpgs.\n",
    "        \"\"\"\n",
    "        raise NotImplementedError\n",
    "    \n",
    "    def tensorize(self):\n",
    "        \"\"\"\n",
    "        Represent sequence as tensor.\n",
    "        \"\"\"\n",
    "        t = seq2tensor(self.promoter_seq)\n",
    "        self.tensor = t.copy()\n",
    "        return t\n",
    "    \n",
    "    def methylate_tensor(self, cpgs):\n",
    "        \"\"\"\n",
    "        Methylate the tensor with values.\n",
    "        TODO: think how to interface Gene - Cpg - Dataset...\n",
    "        \"\"\"\n",
    "        assert self.tensor not None\n",
    "        \n",
    "        for cpg in cpgs:\n",
    "            # insert cpg value in row for meth_base (cytosine)\n",
    "            cpg_loc = cpg[0]\n",
    "            cpg_value = cpg[1]\n",
    "            seq = methylate_seq(self.tensor, cpg_loc, cpg_value, self.meth_idx, self.mask_idx)\n",
    "                    \n",
    "        self.tensor = seq\n",
    "        return seq\n",
    "    \n",
    "    def __str__(self):\n",
    "        return self.name\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return self.tensor\n",
    "    \n",
    "    \n",
    "g = Gene(\"ENS000000\")\n",
    "g.name, g.promoter_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": [
    "'''\n",
    "Key data loading functionality should be implemented in a dataset object and look like this:\n",
    "\n",
    "get_data(projects: list, genes: list):\n",
    "    return (seq, cpgs, expr)\n",
    "    \n",
    "Regarding data flow:\n",
    "1) Create all CpG objects from Illumina file\n",
    "2) Create all gene objects (here define cpg appartenence to gene)\n",
    "'''\n",
    "\n",
    "\n",
    "class Cpg(object):\n",
    "    \"\"\"\n",
    "    .\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, name):\n",
    "        \"\"\"\n",
    "        .\n",
    "        \"\"\"\n",
    "        self.name = name\n",
    "        self.start = None\n",
    "        self.end = None\n",
    "        self.strand = None\n",
    "        self.genes = []\n",
    "\n",
    "        \n",
    "class PromoterDataset(Dataset):\n",
    "    \"\"\"\n",
    "    Model an immutable container for dataset.\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, projects: list, genes: list):\n",
    "        '''Initializes dataset with given projects and genes.'''\n",
    "        \n",
    "        self.projects = []\n",
    "        for proj_name in projects:\n",
    "            # Add each project as Project object, which points to all relevant files for loading records\n",
    "            # in __get__item\n",
    "            self.projects.append(Project(proj_name))\n",
    "        \n",
    "        self.genes = []\n",
    "    \n",
    "    def __len__(self):\n",
    "        return len(self.projects) * len(self.genes)\n",
    "    \n",
    "    def __getitem__(self):\n",
    "        return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": [
    "# import unittest\n",
    "\n",
    "# class TestProject(unittest.TestCase):\n",
    "#     \"\"\".\"\"\"\n",
    "    \n",
    "#     self.case = 'TCGA-2G-AAHC-01A'\n",
    "#     self.genes = ['ENSG00000242268.2', \"ENSG00000167578.15\", \"ENSG00000105063.17\"]\n",
    "#     self.cpgs = ['cg00000029', 'cg00000165', 'rs966367', 'rs9363764']\n",
    "    \n",
    "#     def test_get_case_data(self):\n",
    "#         raise NotImplementedError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": [
    "proj = 'TCGA-TGCT'\n",
    "proj_path = '/data/eugen/tcga/projects/TCGA-TGCT'\n",
    "\n",
    "P = Project(proj, proj_path)\n",
    "\n",
    "case = 'TCGA-2G-AAHC-01A'\n",
    "genes = ['ENSG00000242268.2', \"ENSG00000167578.15\", \"ENSG00000105063.17\"]\n",
    "cpgs = ['cg00000029', 'cg00000165', 'rs966367', 'rs9363764']\n",
    "\n",
    "case_data = P.get_case_data(case, genes, cpgs)\n",
    "expr = case_data['expression']\n",
    "meth = case_data['methylation']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Beta_value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Composite Element REF</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cg00000029</th>\n",
       "      <td>0.374425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000165</th>\n",
       "      <td>0.099877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs966367</th>\n",
       "      <td>0.526726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs9363764</th>\n",
       "      <td>0.487821</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Beta_value\n",
       "Composite Element REF            \n",
       "cg00000029               0.374425\n",
       "cg00000165               0.099877\n",
       "rs966367                 0.526726\n",
       "rs9363764                0.487821"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "#p = Path(\"/data/eugen/tcga/projects\") / proj\n",
    "#[f for f in p.iterdir()]\n",
    "meth = pd.read_csv(Path(\"/data/eugen/tcga/projects\") / proj /\"TCGA-TGCT_methylation.csv\", sep='\\t')\n",
    "expr = pd.read_csv(Path(\"/data/eugen/tcga/projects\") / proj /\"TCGA-TGCT_expression.csv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_name</th>\n",
       "      <th>cases</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cases</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2G-AAHC-01A-11D-A42Z-05</th>\n",
       "      <td>Methylation Beta Value</td>\n",
       "      <td>jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>TCGA-2G-AAHC-01A-11D-A42Z-05</td>\n",
       "      <td>016d83e0-259e-45c4-8bd3-5c784ac459fd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2G-AAGT-01A-11D-A42Z-05</th>\n",
       "      <td>Methylation Beta Value</td>\n",
       "      <td>jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>TCGA-2G-AAGT-01A-11D-A42Z-05</td>\n",
       "      <td>48862052-ac4d-4691-bfb9-d244e5681f90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2G-AALY-01A-11D-A42Z-05</th>\n",
       "      <td>Methylation Beta Value</td>\n",
       "      <td>jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>TCGA-2G-AALY-01A-11D-A42Z-05</td>\n",
       "      <td>23760992-0067-467b-9ee3-843056ff60fb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2G-AAH8-01A-11D-A42Z-05</th>\n",
       "      <td>Methylation Beta Value</td>\n",
       "      <td>jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>TCGA-2G-AAH8-01A-11D-A42Z-05</td>\n",
       "      <td>b454b4f8-1822-4bd4-8c36-752c4fbbd3c8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2G-AAGS-01A-11D-A42Z-05</th>\n",
       "      <td>Methylation Beta Value</td>\n",
       "      <td>jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>TCGA-2G-AAGS-01A-11D-A42Z-05</td>\n",
       "      <td>5c06ef79-0993-43af-9613-c08e82775b35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2G-AAG9-01A-11D-A42Z-05</th>\n",
       "      <td>Methylation Beta Value</td>\n",
       "      <td>jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>TCGA-2G-AAG9-01A-11D-A42Z-05</td>\n",
       "      <td>dd66c667-b77c-4516-93d3-36f576ed5a2c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2G-AAGW-01A-11D-A42Z-05</th>\n",
       "      <td>Methylation Beta Value</td>\n",
       "      <td>jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>TCGA-2G-AAGW-01A-11D-A42Z-05</td>\n",
       "      <td>49296673-6dfb-45ea-8d02-9bee12a0c64c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2G-AAH2-01A-11D-A42Z-05</th>\n",
       "      <td>Methylation Beta Value</td>\n",
       "      <td>jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>TCGA-2G-AAH2-01A-11D-A42Z-05</td>\n",
       "      <td>2af135ce-8d51-465d-bef1-7279bfa3c2ee</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2G-AAGF-01A-31D-A42Z-05</th>\n",
       "      <td>Methylation Beta Value</td>\n",
       "      <td>jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>TCGA-2G-AAGF-01A-31D-A42Z-05</td>\n",
       "      <td>5f761671-91ef-48de-a496-57bacad8306b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-YU-A94I-01A-11D-A436-05</th>\n",
       "      <td>Methylation Beta Value</td>\n",
       "      <td>jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>TCGA-YU-A94I-01A-11D-A436-05</td>\n",
       "      <td>d8501c30-b180-44bc-9a7d-aa3f56183586</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>156 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           data_type  \\\n",
       "cases                                                  \n",
       "TCGA-2G-AAHC-01A-11D-A42Z-05  Methylation Beta Value   \n",
       "TCGA-2G-AAGT-01A-11D-A42Z-05  Methylation Beta Value   \n",
       "TCGA-2G-AALY-01A-11D-A42Z-05  Methylation Beta Value   \n",
       "TCGA-2G-AAH8-01A-11D-A42Z-05  Methylation Beta Value   \n",
       "TCGA-2G-AAGS-01A-11D-A42Z-05  Methylation Beta Value   \n",
       "...                                              ...   \n",
       "TCGA-2G-AAG9-01A-11D-A42Z-05  Methylation Beta Value   \n",
       "TCGA-2G-AAGW-01A-11D-A42Z-05  Methylation Beta Value   \n",
       "TCGA-2G-AAH2-01A-11D-A42Z-05  Methylation Beta Value   \n",
       "TCGA-2G-AAGF-01A-31D-A42Z-05  Methylation Beta Value   \n",
       "TCGA-YU-A94I-01A-11D-A436-05  Methylation Beta Value   \n",
       "\n",
       "                                                                      file_name  \\\n",
       "cases                                                                             \n",
       "TCGA-2G-AAHC-01A-11D-A42Z-05  jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...   \n",
       "TCGA-2G-AAGT-01A-11D-A42Z-05  jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...   \n",
       "TCGA-2G-AALY-01A-11D-A42Z-05  jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...   \n",
       "TCGA-2G-AAH8-01A-11D-A42Z-05  jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...   \n",
       "TCGA-2G-AAGS-01A-11D-A42Z-05  jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...   \n",
       "...                                                                         ...   \n",
       "TCGA-2G-AAG9-01A-11D-A42Z-05  jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...   \n",
       "TCGA-2G-AAGW-01A-11D-A42Z-05  jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...   \n",
       "TCGA-2G-AAH2-01A-11D-A42Z-05  jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...   \n",
       "TCGA-2G-AAGF-01A-31D-A42Z-05  jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...   \n",
       "TCGA-YU-A94I-01A-11D-A436-05  jhu-usc.edu_TGCT.HumanMethylation450.1.lvl-3.T...   \n",
       "\n",
       "                                                     cases  \\\n",
       "cases                                                        \n",
       "TCGA-2G-AAHC-01A-11D-A42Z-05  TCGA-2G-AAHC-01A-11D-A42Z-05   \n",
       "TCGA-2G-AAGT-01A-11D-A42Z-05  TCGA-2G-AAGT-01A-11D-A42Z-05   \n",
       "TCGA-2G-AALY-01A-11D-A42Z-05  TCGA-2G-AALY-01A-11D-A42Z-05   \n",
       "TCGA-2G-AAH8-01A-11D-A42Z-05  TCGA-2G-AAH8-01A-11D-A42Z-05   \n",
       "TCGA-2G-AAGS-01A-11D-A42Z-05  TCGA-2G-AAGS-01A-11D-A42Z-05   \n",
       "...                                                    ...   \n",
       "TCGA-2G-AAG9-01A-11D-A42Z-05  TCGA-2G-AAG9-01A-11D-A42Z-05   \n",
       "TCGA-2G-AAGW-01A-11D-A42Z-05  TCGA-2G-AAGW-01A-11D-A42Z-05   \n",
       "TCGA-2G-AAH2-01A-11D-A42Z-05  TCGA-2G-AAH2-01A-11D-A42Z-05   \n",
       "TCGA-2G-AAGF-01A-31D-A42Z-05  TCGA-2G-AAGF-01A-31D-A42Z-05   \n",
       "TCGA-YU-A94I-01A-11D-A436-05  TCGA-YU-A94I-01A-11D-A436-05   \n",
       "\n",
       "                                                                id  \n",
       "cases                                                               \n",
       "TCGA-2G-AAHC-01A-11D-A42Z-05  016d83e0-259e-45c4-8bd3-5c784ac459fd  \n",
       "TCGA-2G-AAGT-01A-11D-A42Z-05  48862052-ac4d-4691-bfb9-d244e5681f90  \n",
       "TCGA-2G-AALY-01A-11D-A42Z-05  23760992-0067-467b-9ee3-843056ff60fb  \n",
       "TCGA-2G-AAH8-01A-11D-A42Z-05  b454b4f8-1822-4bd4-8c36-752c4fbbd3c8  \n",
       "TCGA-2G-AAGS-01A-11D-A42Z-05  5c06ef79-0993-43af-9613-c08e82775b35  \n",
       "...                                                            ...  \n",
       "TCGA-2G-AAG9-01A-11D-A42Z-05  dd66c667-b77c-4516-93d3-36f576ed5a2c  \n",
       "TCGA-2G-AAGW-01A-11D-A42Z-05  49296673-6dfb-45ea-8d02-9bee12a0c64c  \n",
       "TCGA-2G-AAH2-01A-11D-A42Z-05  2af135ce-8d51-465d-bef1-7279bfa3c2ee  \n",
       "TCGA-2G-AAGF-01A-31D-A42Z-05  5f761671-91ef-48de-a496-57bacad8306b  \n",
       "TCGA-YU-A94I-01A-11D-A436-05  d8501c30-b180-44bc-9a7d-aa3f56183586  \n",
       "\n",
       "[156 rows x 4 columns]"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#meth = meth.set_index(\"file_id\")\n",
    "#meth = meth.to_dict(orient='index')\n",
    "# meth.set_index('cases', drop=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_id</th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_name</th>\n",
       "      <th>cases</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6ef0f618-69e1-4707-904c-9f6f3d7f25cf</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>3e470f27-ee45-4b57-8b84-2ac4b055dcbc.FPKM-UQ.t...</td>\n",
       "      <td>TCGA-2G-AAHC-01A-11R-A430-07</td>\n",
       "      <td>6ef0f618-69e1-4707-904c-9f6f3d7f25cf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>f95eb680-d99e-4145-b2dd-651311f38acc</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>8c06cbb3-f46e-41b8-8217-bb22ac1b471a.FPKM-UQ.t...</td>\n",
       "      <td>TCGA-2G-AAGT-01A-11R-A430-07</td>\n",
       "      <td>f95eb680-d99e-4145-b2dd-651311f38acc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>60a9f839-bc01-4705-8c0a-7a2adf9686f6</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>0e6e268d-d830-40ed-95a7-38f4ae1e7eac.FPKM-UQ.t...</td>\n",
       "      <td>TCGA-2G-AALY-01A-11R-A430-07</td>\n",
       "      <td>60a9f839-bc01-4705-8c0a-7a2adf9686f6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>27e5ccb2-6f5c-4cbc-89e3-28e3f854ce00</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>7868cab1-0300-493b-b910-6208a3c7921a.FPKM-UQ.t...</td>\n",
       "      <td>TCGA-2G-AAH8-01A-11R-A430-07</td>\n",
       "      <td>27e5ccb2-6f5c-4cbc-89e3-28e3f854ce00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4b3b4c23-7974-4095-a491-f43fa9d982a7</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>b6cf2f29-6ea4-422e-bfa6-7bbc3e07e0b5.FPKM-UQ.t...</td>\n",
       "      <td>TCGA-2G-AAGS-01A-11R-A430-07</td>\n",
       "      <td>4b3b4c23-7974-4095-a491-f43fa9d982a7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                file_id                       data_type  \\\n",
       "0  6ef0f618-69e1-4707-904c-9f6f3d7f25cf  Gene Expression Quantification   \n",
       "1  f95eb680-d99e-4145-b2dd-651311f38acc  Gene Expression Quantification   \n",
       "2  60a9f839-bc01-4705-8c0a-7a2adf9686f6  Gene Expression Quantification   \n",
       "3  27e5ccb2-6f5c-4cbc-89e3-28e3f854ce00  Gene Expression Quantification   \n",
       "4  4b3b4c23-7974-4095-a491-f43fa9d982a7  Gene Expression Quantification   \n",
       "\n",
       "                                           file_name  \\\n",
       "0  3e470f27-ee45-4b57-8b84-2ac4b055dcbc.FPKM-UQ.t...   \n",
       "1  8c06cbb3-f46e-41b8-8217-bb22ac1b471a.FPKM-UQ.t...   \n",
       "2  0e6e268d-d830-40ed-95a7-38f4ae1e7eac.FPKM-UQ.t...   \n",
       "3  7868cab1-0300-493b-b910-6208a3c7921a.FPKM-UQ.t...   \n",
       "4  b6cf2f29-6ea4-422e-bfa6-7bbc3e07e0b5.FPKM-UQ.t...   \n",
       "\n",
       "                          cases                                    id  \n",
       "0  TCGA-2G-AAHC-01A-11R-A430-07  6ef0f618-69e1-4707-904c-9f6f3d7f25cf  \n",
       "1  TCGA-2G-AAGT-01A-11R-A430-07  f95eb680-d99e-4145-b2dd-651311f38acc  \n",
       "2  TCGA-2G-AALY-01A-11R-A430-07  60a9f839-bc01-4705-8c0a-7a2adf9686f6  \n",
       "3  TCGA-2G-AAH8-01A-11R-A430-07  27e5ccb2-6f5c-4cbc-89e3-28e3f854ce00  \n",
       "4  TCGA-2G-AAGS-01A-11R-A430-07  4b3b4c23-7974-4095-a491-f43fa9d982a7  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols = ['file_id', 'file_name', 'cases']\n",
    "expr[cols].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "P = Project('TCGA-TGCT', Path(\"/data/eugen/tcga/projects\") / proj )\n",
    "P.samples['expression']\n",
    "\"f95eb680-d99e-4145-b2dd-651311f38acc\" in P.samples['expression']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": [
    "VOCAB = ['A', 'C', 'T', 'G', 'N']\n",
    "BASE2IDX = {\"A\": 0, \"C\": 1, \"T\": 2, \"G\": 3, \"N\": 4}\n",
    "\n",
    "\n",
    "def findall(base: str, seq: str):\n",
    "    idxs = []\n",
    "    for m in re.finditer(base, seq):\n",
    "        idxs.append(m.start(0))\n",
    "    return idxs\n",
    "\n",
    "\n",
    "def random_seq(l: int, vocab: list = VOCAB) -> str:\n",
    "    return \"\".join(np.random.choice(vocab, l))\n",
    "\n",
    "\n",
    "def random_data(m: int, l: int, c: int) -> tuple:\n",
    "    \"\"\"\n",
    "    Generates random dataset consisting of m sequences of length l with up to c cpg sites per sequence.\n",
    "    \"\"\"\n",
    "    assert c < l\n",
    "\n",
    "    seqs = [random_seq(l, ) for _ in range(m)]\n",
    "    cpgs = [[(np.random.choice(findall(\"C\", seqs[i])), np.random.random()) for _ in range(np.random.randint(0, c))] for i in range(m)]\n",
    "    exprs = [float(np.random.binomial(n=1, p=x.count('N')/len(x), size=1)[0]) for x in seqs]\n",
    "\n",
    "    return seqs, cpgs, exprs\n",
    "\n",
    "\n",
    "def base2tensor(base: str, vocab: list = VOCAB) -> torch.tensor:\n",
    "    tensor = torch.zeros(1, len(vocab))\n",
    "    tensor[BASE2IDX[base]][0] = 1\n",
    "    return tensor\n",
    "\n",
    "\n",
    "def seq2tensor(seq: str, vocab: list = VOCAB) -> torch.tensor:\n",
    "    tensor = torch.zeros(len(seq), 1, len(vocab))  # important choice for preserving shape compatibility with hidden layer.\n",
    "    for idx, base in enumerate(seq):\n",
    "        tensor[idx][0][BASE2IDX[base]] = 1  # that extra 1 dimension is because PyTorch assumes everything is in batches - we’re just using a batch size of 1 here.\n",
    "    return tensor\n",
    "\n",
    "\n",
    "def methylate_seq(seq_tensor: torch.tensor, loc: int, value: float, meth_idx: int, mask_idx: int) -> torch.tensor:\n",
    "    assert (seq_tensor[loc][0][meth_idx] > 0) or (seq_tensor[loc][0][mask_idx] > 0)# assert it is a \"C\" in first place\n",
    "    seq_tensor[loc][0][meth_idx] = value\n",
    "    return seq_tensor\n",
    "\n",
    "\n",
    "class PromoterArtificialDataset(Dataset):\n",
    "    \"\"\"\n",
    "    Dataset class for loading data. Communicates with loader object to bring batched data.\n",
    "    \n",
    "    @attributes\n",
    "    @methods\n",
    "    \n",
    "    Example:\n",
    "    ...\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, data: tuple):\n",
    "        \n",
    "        self.seqs, self.cpgs, self.exprs = data\n",
    "        \n",
    "        # get some utile attributes\n",
    "        self.vocab = sorted(set(\"\".join(self.seqs)))  # returns list\n",
    "        self.vocab_size = len(self.vocab)\n",
    "        self.base_to_idx = {}\n",
    "        for i, base in enumerate(VOCAB):\n",
    "            self.base_to_idx[base] = i\n",
    "        \n",
    "        self.meth_base = \"C\"  #set methylated base\n",
    "        self.mask_base = \"N\"\n",
    "        self.meth_idx = self.base_to_idx[self.meth_base]\n",
    "        self.mask_idx = self.base_to_idx[self.mask_base]\n",
    "        \n",
    "        \n",
    "    def __len__(self):\n",
    "        return len(self.exprs)\n",
    "\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        \"\"\"Getter.\"\"\"\n",
    "        seq = seq2tensor(self.seqs[idx])\n",
    "        cpgs = self.cpgs[idx]\n",
    "        expr = self.exprs[idx]\n",
    "\n",
    "        for cpg in cpgs:\n",
    "            # insert cpg value in row for meth_base (cytosine)\n",
    "            cpg_loc = cpg[0]\n",
    "            cpg_value = cpg[1]\n",
    "            seq = methylate_seq(seq, cpg_loc, cpg_value, self.meth_idx, self.mask_idx)\n",
    "\n",
    "        return seq, expr\n",
    "            \n",
    "\n",
    "m = 100\n",
    "l = 50\n",
    "c = 5\n",
    "VOCAB = ['A', 'C', 'T', 'G', 'N']\n",
    "BASE2IDX = {\"A\": 0, \"C\": 1, \"T\": 2, \"G\": 3, \"N\": 4}\n",
    "data = random_data(m, l, c)\n",
    "\n",
    "# test_loader, train_loader\n",
    "train_set = PromoterArtificialDataset(data)\n",
    "train_loader = DataLoader(train_set, batch_size=4, shuffle=True, num_workers=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": [
    "class RNNRegressor(nn.Module):\n",
    "\n",
    "    def __init__(self, input_dim, hidden_dim, output_dim):\n",
    "        super(RNNRegressor, self).__init__()\n",
    "        self.input_dim = input_dim\n",
    "        self.hidden_dim = hidden_dim\n",
    "        self.output_dim = output_dim\n",
    "        self.i2h = nn.Linear(input_dim + hidden_dim, hidden_dim)\n",
    "        self.i2o = nn.Linear(input_dim + hidden_dim, output_dim)\n",
    "        self.tanh = nn.Tanh()\n",
    "    \n",
    "    def forward(self, input, hidden):\n",
    "        \n",
    "        #input = input.view(-1, input.shape[1])\n",
    "        # TODO: Modify code to accept batch tensors <loc, batch_nr, index>. Can add init hidden in here.\n",
    "        \n",
    "        combined = torch.cat((input, hidden), 1)\n",
    "        hidden = self.i2h(combined)\n",
    "        output = self.i2o(combined)\n",
    "        output = self.tanh(output)\n",
    "\n",
    "        return output, hidden\n",
    "    \n",
    "    def initHidden(self):\n",
    "        return torch.zeros(1, self.hidden_dim)\n",
    "\n",
    "HIDDEN_DIM = 10\n",
    "rnn = RNNRegressor(train_set.vocab_size, HIDDEN_DIM, 1)\n",
    "\n",
    "\n",
    "\n",
    "writer = SummaryWriter('runs/RNN_playground')\n",
    "writer.add_graph(rnn, (seq2tensor(\"ACGTT\")[1], torch.zeros(1, HIDDEN_DIM)))\n",
    "writer.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "torch.Size([5, 1, 5]) torch.Size([1])\n"
     ]
    }
   ],
   "source": [
    "criterion = nn.MSELoss()\n",
    "optimizer = optim.SGD(rnn.parameters(), lr=0.005)\n",
    "\n",
    "\n",
    "def train(output_tensor, seq_tensor):\n",
    "    ''''''\n",
    "    \n",
    "    hidden = rnn.initHidden()\n",
    "    \n",
    "    # zero the grad buffers\n",
    "    rnn.zero_grad()\n",
    "    optimizer.zero_grad()\n",
    "    \n",
    "    # forward pass\n",
    "    for i in range(seq_tensor.shape[0]):\n",
    "        output, hidden = rnn(seq_tensor[i], hidden)\n",
    "    \n",
    "    # compute loss and backward pass\n",
    "    loss = criterion(output, output_tensor)\n",
    "    loss.backward()\n",
    "    \n",
    "    # update params\n",
    "    optimizer.step()\n",
    "    \n",
    "    return output, loss.item()\n",
    "\n",
    "train(torch.tensor([-0.0117]), seq2tensor(\"ACGTN\"))\n",
    "print(seq2tensor(\"ACGTN\").shape, torch.tensor([-0.0118]).shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished Training\n"
     ]
    }
   ],
   "source": [
    "NUM_EPOCHS = 5\n",
    "\n",
    "# TODO: not really minibatch for now\n",
    "for epoch in range(NUM_EPOCHS):\n",
    "    \n",
    "    running_loss = 0.0\n",
    "    for i, data in enumerate(train_loader, 0):\n",
    "        \n",
    "        # get inputs\n",
    "        # TODO: Account for dtypes, otherwise get incompatible!\n",
    "        seq_tensor, expr = data\n",
    "        seq_tensor = seq_tensor[0]\n",
    "        expr = expr[0].view(1)\n",
    "        expr = expr.type(torch.FloatTensor)  \n",
    "        \n",
    "        # train on example\n",
    "        output, loss = train(expr[0], seq_tensor)\n",
    "        \n",
    "        # print statistics\n",
    "        running_loss += loss\n",
    "        if i % 2000 == 1999:    # print every 2000 mini-batches\n",
    "            print('[%d, %5d] loss: %.3f' %\n",
    "                  (epoch + 1, i + 1, running_loss / 2000))\n",
    "            running_loss = 0.0\n",
    "\n",
    "print('Finished Training')\n",
    "\n",
    "# save model\n",
    "PATH = './models/test_model'\n",
    "torch.save(rnn.state_dict(), PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": [
    "# Predict loop\n",
    "\n",
    "# correct = 0\n",
    "# total = 0\n",
    "# with torch.no_grad():\n",
    "#     for data in testloader:\n",
    "#         images, labels = data\n",
    "#         outputs = net(images)\n",
    "#         _, predicted = torch.max(outputs.data, 1)\n",
    "#         total += labels.size(0)\n",
    "#         correct += (predicted == labels).sum().item()\n",
    "\n",
    "# print('Accuracy of the network on the 10000 test images: %d %%' % (\n",
    "#     100 * correct / total))"
   ]
  }
 ],
 "metadata": {
  "file_extension": ".py",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  },
  "mimetype": "text/x-python",
  "name": "python",
  "npconvert_exporter": "python",
  "pygments_lexer": "ipython3",
  "version": 3
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
